mono/metadata/class-init.c

```diff
@@ -1612,15 +1612,10 @@ set_interface_and_offset (int num_ifaces, MonoClass **interfaces_full, int *inte
 		return FALSE;
 	}
 	for (i = 0; i < num_ifaces; ++i) {
-		if (interfaces_full [i]) {
-			int end;
-			if (interfaces_full [i]->interface_id < ic->interface_id)
-				continue;
-			end = i + 1;
-			while (end < num_ifaces && interfaces_full [end]) end++;
-			memmove (interfaces_full + i + 1, interfaces_full + i, sizeof (MonoClass*) * (end - i));
-			memmove (interface_offsets_full + i + 1, interface_offsets_full + i, sizeof (int) * (end - i));
-		}
+		if (interfaces_full [i] && interfaces_full [i]->interface_id == ic->interface_id)
+			return TRUE;
```

Comment:
So it looks like `interfaces_full` isn't stored in sorted order based on `interface_id` anymore.  Because of that the call to `find_interface` is failing (because `find_interface` is doing a binary search based on `interface_id`), and so as a result you need to possibly `return TRUE` here if you find the interface using this linear scan.

Questions:
1. Shouldn't we respect `force_set` here too?
2. Why is it okay to keep `interfaces_full` in unsorted order?
3. I don't understand how this is solving the problems in twopassvariance.il.  Could you explain how the interfaces were laid out before and after this change? (It would be good to include that in the commit message)
4. `interfaces_full` is published to `interfaces_packed` and `interface_offsets_packed` - do all the places where those fields are used work with non-sorted data? 
5. We should get some profiling measurements before/after this change on some interface-heavy example to see if going to a linear search here is bad for us.

Reply:
1) Yes, you are right we have to change find_interface and keep the force_set working. I did a new PR with this.
2) Because we need to follow the definition order to check possibility to cast and other things, we cannot follow the order of ids, because if we do this a class that implements A and B, behave the same way as a class that implements B and A, because the interfaces A and B have the same id in both cases, so they will be ordered in the same way.
3) I wrote on PR message. I don't know if it was this that you want to know. But to explain more, before the changes on IlAsm the order of interfaces were the inverse what was written on source code. And the order of definition of interfaces, depend on the usage of interfaces and not the definition. With the changes, we follow exactly what is on source code. As we use the order of the interfaces_packed to check if the cast is possible, as we were ordered by id, Fooer1 and Fooer2 for example, have the same behavior. 
4) As I could see and understand yes, all of them should work with non-sorted data.
5) Should I write a heavy example or do ou have this anywhere? I think that we would lose a lot of performance if a class implements a lot of interfaces, and I don't know if this is really used. Is a good idea to check from how many interfaces implemented we start loosing performance?

Reply:
About 3. Sorry, I wasn't very clear: I was more interested in an example of how the vtable for a class would look before and after these changes, not the table layout generated by ILasm.  Just a small example of a single class with multiple interfaces that shows that before an incorrect method was used and after it is correct.   Something like the commit message on #10931.

About 5.  That's true a linear search on a small number of interfaces would do alright.

One other new thought: take a look at https://github.com/mono/mono/blob/02946b7cedb79c01da04d200564d864a17659d42/mono/metadata/class-init.c#L1825-L1834  this is the case for generic instances of a class.  In this case we re-shuffle the interfaces into interface_id order again.  So I think a variant of twopassvariance.il where some of the `Fooer` classes are generic will give incorrect answers.

Reply:
3) This is how the Vtable for Fooer1 and Fooer2 looks like after the changes:
*** Vtable for class 'Fooer1' at "FINALLY" (size 6)
  [O][000][INDEX 000] object:ToString () [0x7fd78681db20]
  [O][001][INDEX 001] object:GetHashCode () [0x7fd78681daf8]
  [O][002][INDEX 002] object:Finalize () [0x7fd78681dad0]
  [O][003][INDEX 003] object:Equals (object) [0x7fd78681daa8]
  [I][004][INDEX 000] IFoo<Base>:Gimme () [0x7fd786503800]
  [I][005][INDEX 000] IFoo<Derived>:Gimme () [0x7fd7865038c0]
Packed interface table for class Fooer1 has size 2
  [000][UUID 078][SLOT 004][SIZE  001] interface IFoo[Base]
  [001][UUID 079][SLOT 005][SIZE  001] interface IFoo[Derived]
*** Vtable for class 'Fooer2' at "FINALLY" (size 6)
  [O][000][INDEX 000] object:ToString () [0x7fd78681db20]
  [O][001][INDEX 001] object:GetHashCode () [0x7fd78681daf8]
  [O][002][INDEX 002] object:Finalize () [0x7fd78681dad0]
  [O][003][INDEX 003] object:Equals (object) [0x7fd78681daa8]
  [I][004][INDEX 000] IFoo<Derived>:Gimme () [0x7fd7865038c0]
  [I][005][INDEX 000] IFoo<Base>:Gimme () [0x7fd786503800]
Packed interface table for class Fooer2 has size 2
  [000][UUID 079][SLOT 004][SIZE  001] interface IFoo[Derived]
  [001][UUID 078][SLOT 005][SIZE  001] interface IFoo[Base]

This is how the VTable for Fooer1 and Fooer2 looks like before the changes, as you can see Packed interface table are the same in Fooer1 and Fooer2, so the behavior of both classes was the same:
Vtable for class 'Fooer1' at "FINALLY" (size 6)
  [O][000][INDEX 000] object:ToString () [0x7fcf3801d920]
  [O][001][INDEX 001] object:GetHashCode () [0x7fcf3801d8f8]
  [O][002][INDEX 002] object:Finalize () [0x7fcf3801d8d0]
  [O][003][INDEX 003] object:Equals (object) [0x7fcf3801d8a8]
  [I][004][INDEX 000] IFoo<Base>:Gimme () [0x7fcf37c29880]
  [I][005][INDEX 000] IFoo<Derived>:Gimme () [0x7fcf37c29940]
Packed interface table for class Fooer1 has size 2
  [000][UUID 076][SLOT 004][SIZE  001] interface IFoo[Base]
  [001][UUID 077][SLOT 005][SIZE  001] interface IFoo[Derived]

Vtable for class 'Fooer2' at "FINALLY" (size 6)
  [O][000][INDEX 000] object:ToString () [0x7fcf3801d920]
  [O][001][INDEX 001] object:GetHashCode () [0x7fcf3801d8f8]
  [O][002][INDEX 002] object:Finalize () [0x7fcf3801d8d0]
  [O][003][INDEX 003] object:Equals (object) [0x7fcf3801d8a8]
  [I][004][INDEX 000] IFoo<Derived>:Gimme () [0x7fcf37c29940]
  [I][005][INDEX 000] IFoo<Base>:Gimme () [0x7fcf37c29880]
Packed interface table for class Fooer2 has size 2
  [000][UUID 076][SLOT 005][SIZE  001] interface IFoo[Base]
  [001][UUID 077][SLOT 004][SIZE  001] interface IFoo[Derived]

5) I'll try to create another test that will pass in this part of source code.

Reply:
5) You were right, when I create a new test I could reproduce and change this part of code. I did a PR with the new test and the fix.